### ISSUE TITLE
nrjavaserial 3.12.0 requires glibc > 2.13

### ISSUE THREAD

**Issue by tsheldon** [openhab/openhab-core#111]
tl;dr;  fresh install of oh 2.0.0~20170111215718 which includes nrjavaserial 3.12.0 causes silent failure with zwave (and possibly rfxcom) binding.

After seeing the announcement of moving to Bintray, I went and ran the upgrade as described in the thread.   The system immediately started crashing on startup when trying to open the serial port.  After debugging for a bit, I decided to purge the install and start fresh.   When I did, OH started fine. *until* I installed just the zwave binding.   As soon as the binding tried to open the serial port, it started crashing again.

The only debug info I have been able to retrieve is this:
```
09:19:22.535 [INFO ] [ing.zwave.handler.ZWaveSerialHandler] - Connecting to serial port '/dev/ttyACM0'
/usr/bin/java: symbol lookup error: /var/lib/openhab2/tmp/libNRJavaSerial_HF_openhab_0/libNRJavaSerial_HF.so: undefined symbol: __fdelt_chk
```

**Comment by kaikreuzer** [comment]
@cvanorman Any idea which change between 3.9.3 and 3.12.0 might cause this? Would we have to revert some further commit from the 3.12.0 release?

**Comment by tsheldon** [comment]
So I may have a lead on this-   my pi3 runs on Raspbian GNU/Linux 7 (Wheezy).    Looks like the latest compile for nrjavaserial is compiled with a version glibc which is newer than the latest available on Wheezy.

I'm attempting to update to Jessie which should have a newer glibc, but may not be able to complete that until I get home this evening.

**Comment by tsheldon** [comment]
Confirmed.   At some point between the glibc 2.13 of Wheezy and glibc 2.19 of Jessie, fdelt_chk was added.

As it stands, the snapshot build of OH2 is *not* compatible with Raspbian Wheezy.

**Comment by kaikreuzer** [comment]
@cvanorman Do you have a chance to re-compile the binaries against an older glibc?

**Comment by cvanorman** [comment]
I will try and look into that tonight.

**Comment by cvanorman** [comment]
The minimum version currently required would be glibc 2.15.  Is there any general guidance on supported platforms?  I'm not sure what the commonly used platforms for OH are.

I don't have the resources right now to set up more build platforms (I think I'd have to go back to like ubuntu 10.0.4 to get glibc 2.13). I'm not sure what the original repository was using to perform the builds.

Another possibility is to disable the FORTIFY_SOURCE option.  I tried this and it appears to have removed the fdelt_chk symbol dependency.  I've attached a version built with this option (the ARM libraries only):

[nrjavaserial-3.12.0-OH.zip](https://github.com/openhab/openhab-core/files/703484/nrjavaserial-3.12.0-OH.zip)

 if you'd like to test that @tsheldon?

**Comment by eleson2** [comment]
Jessie was first released 25th of April 2015. Which means that many , or most, OH1-RBPi users run on Wheezy.
What is the correct way forward here? 
Stick to supporting Wheezy, or be clear that a upgrade from OH1 requires a OS upgrade / fresh install? 
( I'm assuming that zwave support is essential to the package)

**Comment by cvanorman** [comment]
I provided an attachment just above that will, hopefully, run in Wheezy.  The goal of this particular issue is to resolve that.

**Comment by kaikreuzer** [comment]
If anybody could test it and confirm that it resolves the issue, I can include it in the distro.

**Comment by eleson2** [comment]
I tried to test but no change. But I'm to inexperienced in OH and my setup is far from a controlled environment for a proper test, so don't draw any big conclusions from that.

**Comment by tsheldon** [comment]
I managed to get Wheezy up in a qemu instance, but I've yet to be able to
plumb a serial device through to it to test the recompiled nrjavaserial.

**Comment by kaikreuzer** [comment]
Would be great if anyone could test and confirm soon that this fixes the issue, so that we can include the fix in the final 2.0 release.

@cvanorman Normally, it is not a good practice to replace releases in the repo with a different build, but the very same version - it might lead to problems as it is not always clear, which version of the lib has been picked up by which build.
But as this is based on the very same source code as before and it should actually be fully compatible for everyone where it is working already, I'd be fine with keeping the version and replacing it in the repo. I just wanted to highlight the associated risk.

**Comment by eleson2** [comment]
as a first step, this testing actually doesn't need a zwave device. 
I'm fairly sure this fault occurs before the attached device is ever reached.
But you need to be sure you tested the right stuff. - And here I failed...

**Comment by eleson2** [comment]
Is this implemented in the RC1 ?

**Comment by kaikreuzer** [comment]
No, I am still waiting for anyone to confirm it's working. In any case, it seems to be to risky to change the lib to an untested one on the very last day before packaging the release...

**Comment by tsheldon** [comment]
Finally managed to get qemu with a straight wheezy image running.     Loaded the zwave binding and as soon as I configured the serial port, java bailed out once it attempted to open the serial port.

I did manage to find the java exception from that one which is attached.

Unfortunately I had the same result after overwriting the existing nrjavaserial-3.12.0-OH.jar with the unzipped version which @cvanorman attached.   I've attached that crash as well.

nm still reports __fdelt_chk as undefined:
repository jar:
```
$ nm -n libNRJavaSerial_HF.so | grep __fdelt_chk
         U __fdelt_chk
```

jar attached by @cvanorman 
```
$ nm -n libNRJavaSerial_HF.so | grep __fdelt_chk
         U __fdelt_chk
```

Infact, there's no change in the number of undefined symbols between the two libraries.

**Comment by tsheldon** [comment]
I guess it helps to attach those logs....
[java_crash_preupdate.txt](https://github.com/openhab/openhab-core/files/720610/java_crash_preupdate.txt)
[java_crash_postupdate.txt](https://github.com/openhab/openhab-core/files/720611/java_crash_postupdate.txt)

**Comment by tsheldon** [comment]
I was going to attempt to rebuild nrjavaserial on the my emulated pi, but looks like the latest version of gradle included in wheezy is 1.8.6 which appears too old to build the library.

**Comment by mschulte** [comment]
hey @cvanorman, @kaikreuzer,

i can confirm that the patched [nrjavaserial-3.12.0-OH.zip](https://github.com/openhab/openhab-core/files/703484/nrjavaserial-3.12.0-OH.zip) from above works fine for me.

i got the "symbol lookup error" using CUL Transport together with FS20 and FHT Binding after updating from beta5 to RC1.
After removing the serial bundle and extracting the above mentioned jar to my "addons" folder my FS20 and FHT Devices work well again.

You may include that patch in the upcoming release?

**Comment by cvanorman** [comment]
@tsheldon  Are you sure you used the correct JAR file? I just downloaded it and pulled out the native library and I get no reference to __fdelt_chk.

**Comment by conditio** [comment]
Sorry to disturb, but having a hard time migrating from OH1 on Centos 6.5 to the openhab 2.0 official release it gives this error:
java.lang.UnsatisfiedLinkError: /usr/local/openhab2/userdata/tmp/libNRJavaSerial_root_0/libNRJavaSerial.so: /lib/libc.so.6: version `GLIBC_2.15' not found (required by /usr/local/openhab2/userdata/tmp/libNRJavaSerial_root_0/libNRJavaSerial.so)
Is this related to the issue you're discussing?
How can this be solved....

**Comment by mschulte** [comment]
yes it is.

as a workaround, remove the serial binding from your openhab installation. then take the patched https://github.com/openhab/openhab-core/files/703484/nrjavaserial-3.12.0-OH.zip from above and put it in your addons folder and it should work.

**Comment by conditio** [comment]
Where is the serial binding in OH2?
I can't see it anywhere
in addons.cfg I only have:
binding = zwave,sonos,dsc,netatmo

MVH
Johan Lund, CondITio
0707-15 81 36

**Comment by cvanorman** [comment]
The patched file referenced here only corrects the glibc dependency for ARM hardware.  It will not work on a CentOS 6 system.

**Comment by conditio** [comment]
Ok then its no use for me, how can I get around this? I guess I'm not the only one who's about to migrate from OH1 on a Centos 6.X installation...

**Comment by conditio** [comment]
Is this something thats in working progress, maybe I just postpone my migration a couple of weeks/months or whats the timeframe? Can I assist in some way?

**Comment by cvanorman** [comment]
OK, I've applied the same fix to the 32-bit and 64-bit linux natives.


[nrjavaserial-3.12.1-OH.zip](https://github.com/openhab/openhab-core/files/768901/nrjavaserial-3.12.1-OH.zip)

Give it a try.

**Comment by conditio** [comment]
I've given it a try...
Don't know if I've done it correct.
First put the jar (from inside the zip) in ./openhab2/addons/nrjavaserial-3.12.1-OH.jar

then removed the ./openhab2/runtime/system/org/openhab/nrjavaserial/3.12.0.OH/nrjavaserial-3.12.0.OH.jar
away

tried again same error as before:
java.lang.UnsatisfiedLinkError: /usr/local/openhab2/userdata/tmp/libNRJavaSerial_root_0/libNRJavaSerial.so: /lib/libc.so.6: version `GLIBC_2.15' not found (required by /usr/local/openhab2/userdata/tmp/libNRJavaSerial_root_0/libNRJavaSerial.so)

Then I removed all the files  libNRJavaSerial_* within userdata/tmp since they didn't seem to have changed.
Restarted and the same error as above again...

Am I doing something wrong or is there still something which requires 2.15 within the jar...

**Comment by conditio** [comment]
Don't know if it matters, but maybe I should tell you that I'm trying to reference the z-wave stick through a symlink "/dev/USBzwave" which has been working fine in my openhab 1 installation...
https://github.com/openhab/openhab1-addons/wiki/symlinks

**Comment by conditio** [comment]
Nothing new, I'm really up for migrating, but seems a bit too much work to update to CentOS 7...

**Comment by conditio** [comment]
still no news...

**Comment by conditio** [comment]
Stil having this problem, both with 2.1 Release and with the  2.2.0-SNAPSHOT Build #988 on CentOS 6.5
Any sugestions, upgrading to other OS is not an option...

openhab> java.lang.UnsatisfiedLinkError: /usr/local/openhab-2.1.0/userdata/tmp/libNRJavaSerial_root_0/libNRJavaSerial.so: /lib/libc.so.6: version `GLIBC_2.15' not found (required by /usr/local/openhab-2.1.0/userdata/tmp/libNRJavaSerial_root_0/libNRJavaSerial.so)
        at java.lang.ClassLoader$NativeLibrary.load(Native Method)
        at java.lang.ClassLoader.loadLibrary0(ClassLoader.java:1941)
        at java.lang.ClassLoader.loadLibrary(ClassLoader.java:1824)
        at java.lang.Runtime.load0(Runtime.java:809)
        at java.lang.System.load(System.java:1086)
        at gnu.io.NativeResource.loadResource(NativeResource.java:142)
        at gnu.io.NativeResource.inJarLoad(NativeResource.java:40)
        at gnu.io.NativeResource.loadLib(NativeResource.java:60)
        at gnu.io.NativeResource.load(NativeResource.java:28)
        at gnu.io.SerialManager.<init>(SerialManager.java:10)
        at gnu.io.SerialManager.getInstance(SerialManager.java:16)
        at gnu.io.RXTXCommDriver.<clinit>(RXTXCommDriver.java:87)
        at java.lang.Class.forName0(Native Method)
        at java.lang.Class.forName(Class.java:264)
        at gnu.io.CommPortIdentifier.<clinit>(CommPortIdentifier.java:109)
        at org.openhab.binding.zwave.handler.ZWaveSerialHandler.initialize(ZWaveSerialHandler.java:81)
        at org.eclipse.smarthome.core.thing.internal.ThingManager$8.call(ThingManager.java:731)
        at org.eclipse.smarthome.core.thing.internal.ThingManager$8.call(ThingManager.java:1)
        at org.eclipse.smarthome.core.common.SafeMethodCaller$CallableWrapper.call(SafeMethodCaller.java:181)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:745)
java.lang.UnsatisfiedLinkError: /usr/local/openhab-2.1.0/userdata/tmp/libNRJavaSerial_root_0/libNRJavaSerial.so: /lib/libc.so.6: version `GLIBC_2.15' not found (required by /usr/local/openhab-2.1.0/userdata/tmp/libNRJavaSerial_root_0/libNRJavaSerial.so)
        at java.lang.ClassLoader$NativeLibrary.load(Native Method)
        at java.lang.ClassLoader.loadLibrary0(ClassLoader.java:1941)
        at java.lang.ClassLoader.loadLibrary(ClassLoader.java:1824)
        at java.lang.Runtime.load0(Runtime.java:809)
        at java.lang.System.load(System.java:1086)
        at gnu.io.NativeResource.loadResource(NativeResource.java:142)
        at gnu.io.NativeResource.inJarLoad(NativeResource.java:40)
        at gnu.io.NativeResource.loadLib(NativeResource.java:60)
        at gnu.io.NativeResource.load(NativeResource.java:28)
        at gnu.io.SerialManager.<init>(SerialManager.java:10)
        at gnu.io.SerialManager.getInstance(SerialManager.java:16)
        at gnu.io.RXTXCommDriver.<clinit>(RXTXCommDriver.java:87)
        at java.lang.Class.forName0(Native Method)
        at java.lang.Class.forName(Class.java:264)
        at gnu.io.CommPortIdentifier.<clinit>(CommPortIdentifier.java:109)
        at org.openhab.binding.zwave.handler.ZWaveSerialHandler.initialize(ZWaveSerialHandler.java:81)
        at org.eclipse.smarthome.core.thing.internal.ThingManager$8.call(ThingManager.java:731)
        at org.eclipse.smarthome.core.thing.internal.ThingManager$8.call(ThingManager.java:1)
        at org.eclipse.smarthome.core.common.SafeMethodCaller$CallableWrapper.call(SafeMethodCaller.java:181)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:745)
java.lang.ExceptionInInitializerError thrown while loading gnu.io.RXTXCommDriver
java.lang.NoClassDefFoundError: Could not initialize class gnu.io.RXTXCommDriver thrown while loading gnu.io.RXTXCommDriver

**Comment by ulfwin** [comment]
I'm having exactly the same issue as originally posted (undefined symbol). I'm using debian 7 (glibc 2.13) on an ARM board and openhab 2.1.0. 

I tried to replace nrjavaserial with the patched one in this post made by @cvanorman , but I must be doing something wrong. I replaced the file in this directory:

/usr/share/openhab2/runtime/system/org/openhab/nrjavaserial/3.12.0.OH

It didn’t change anything, but also I noticed that I could remove the file completely with exactly the same behavior (I would expect some openhab error message instead). Is the addon actually used located somewhere else perhaps?

Or have I misunderstood it and the only way out is to upgrade the OS?

**Comment by ulfwin** [comment]
With the kind help of people in the openhab forum I got it working (see https://community.openhab.org/t/solved-z-wave-dongle-makes-openhab-crash). 

Here are the steps I took:

Make sure openhab is running, for example by removing the dongle and starting the service. Then enter the Karaf consol with (default password is "habopen"):
ssh -p 8101 openhab@localhost

Uninstall the addon:
bundle:uninstall nrjavaserial

Exit Karaf with "logout", and stop openhab:
sudo service openhab2 stop

Download and unzip the addon (this is for ARM systems, choose the other link above if yours is not):
wget https://github.com/openhab/openhab-core/files/703484/nrjavaserial-3.12.0-OH.zip
unzip nrjavaserial-3.12.0-OH.zip

Move the file to the addons folder:
sudo mv nrjavaserial-3.12.0-OH.jar /usr/share/openhab2/addons/

Start the service:
sudo service openhab2 start

It should now work! Plug the dongle in and check it out in paperUI, it should be listed as online.

Hope this can be helpful for someone.

**Comment by kaikreuzer** [comment]
Glad you were able to solve your issue!
Just for me to understand: What is the linked nrjavaserial-3.12.0-OH.zip and how was it built?
Why should there be a reason to require different jars for different OSes? The main idea of nrjavaserial is to bundle the right versions for the different platforms...

**Comment by ulfwin** [comment]
I'll try to answer as well as I can! My part was simply asking for help and using other people's work. 

The link is taken from a previous comment in this thread, by @cvanorman. From his comment I know it was built for ARM, but that's all I know I'm afraid. 

My interpretation is that using the special jar file is a work-around until this issue get solved. Perhaps solving it has been down prioritized because it only afflicts people with older Linux OSes?

**Comment by kaikreuzer** [comment]
Thanks. So maybe @cvanorman can shed some light on it?

> until this issue get solved

Is there at all an issue at https://github.com/NeuronRobotics/nrjavaserial about that? We should try whether the official 3.13.0 release has any such issue or not as the plan is to switch back to an official release (@cvanorman's fork was only a temporary solution).

**Comment by conditio** [comment]
Wonderful this worked for me also!
Did never realize that I had to uninstall through karaf.
The 64 bit version of the file which cvanorman supplied me with ages ago actually worked, but I never got it "installed" the right way...
Thank you all!
PS.
It would be great if this would work out of the box...

**Comment by ulfwin** [comment]
@kaikreuzer: I'm not aware of any reported issue there. I did a search, and only found it being mentioned here:
https://github.com/NeuronRobotics/nrjavaserial/issues/91

It was not reported as a separate issue because the reporter chose to update to newer OS. 

Do you expect that the latest official release can work with OH2 without modification? If so I don't mind trying it (in 2 weeks or so probably, I'm traveling now).

**Comment by eleson2** [comment]
Great that this works!
I bough myself a new memory card and did a fresh install of Jessie and got around it.
( Fresh install was recommended , but it is a pain.)

Good info that karaf should be used for uninstall, that wasn't obvious to me.

**Comment by kaikreuzer** [comment]
> Do you expect that the latest official release can work with OH2 without modification? 

I don't know, but it should be our goal. So if possible, please help testing and improving the official lib, so that we can include it again one day in an unmodified version!

**Comment by bits4life** [comment]
I have the same issue on my dockerized openhab/openhab:2.2.0-arm64-debian install. I tried to use @ulfwin sugestion to replace the bundle. Karaf reports the followoing to me: 
14:13:05.540 [WARN ] [org.apache.felix.fileinstall         ] - A bundle with the same symbolic name (com.neuronrobotics.nrjavaserial) and version (3.12.0.OH) is already installed.  Updating this bundle instead.

Then when using paperUI to create rfxcom tcp/ip thing openhab crashes :
/jni/linux/arm-hf/libjd2xx.so
java: pthread_mutex_lock.c:149: __pthread_mutex_lock: Assertion `PTHREAD_MUTEX_TYPE (mutex) == PTHREAD_MUTEX_ERRORCHECK_NP' failed.
Aborted (core dumped)

Another crash when trying to fix things gave me to following output :
openhab> /jni/linux/arm-hf/libjd2xx.so                                                                                                                            
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0xd9668508, pid=157, tid=0xd93ff470
#
# JRE version: OpenJDK Runtime Environment (8.0_131-b57) (build 1.8.0_131-b57)
# Java VM: OpenJDK Client VM (25.131-b57 mixed mode, Evaluation linux-aarch32 )
libusbx: debug [libusb_get_device_list] 
# Problematic frame:
# [thread -596167568 also had an error]
C  [libjd2xx3415278496678108151.so+0x32508]  libusb_get_next_timeout+0xb8
#
# Core dump written. Default location: /openhab/userdata/core or core.157
#
# An error report file with more information is saved as:
# /openhab/userdata/hs_err_pid157.log
#
# If you would like to submit a bug report, please visit:
#   http://www.azulsystems.com/support/
#
Aborted (core dumped)

Any idea what I can do try to make this work.
Thanks already.
Nico

**Comment by maggu2810** [comment]
Do we still need to support something older then Debian Stretch in 2019?

**Comment by conditio** [comment]
I still run openhab on the same centos 6.5 machine, but thats me...

**Comment by wborn** [comment]
If you're still running an OS with GLIBC < 2.15 in 2020, I suggest you upgrade your OS to something more recent. According to the [life-cycle dates](https://access.redhat.com/support/policy/updates/errata#Life_Cycle_Dates) CentOS 6 will no longer receive maintenance support on November 30, 2020. GLIBC 2.15 was [released](https://abi-laboratory.pro/?view=timeline&l=glibc) in 2012.

With https://github.com/openhab/openhab-core/pull/1426 we will be using the official nrjavaserial (3.20.0) release again. So any related issues should be created in the nrjavaserial [issue tracker](https://github.com/NeuronRobotics/nrjavaserial/issues).

The official nrjavaserial native libraries also require GLIBC 2.15:

```
$ strings libNRJavaSerial.so | grep GLIBC | sed -E 's#.+@@GLIBC(.+)#GLIBC\1#' | sort --unique --version-sort 
GLIBC_2.0
GLIBC_2.1
GLIBC_2.1.3
GLIBC_2.3.4
GLIBC_2.4
GLIBC_2.7
GLIBC_2.15
```

You can check your GLIBC version with `ldd -version`

Ubuntu 12.04:

```
 $ ldd --version
ldd (Ubuntu EGLIBC 2.15-0ubuntu10.13) 2.15
```

Ubuntu 18.04:

```
$ ldd --version
ldd (Ubuntu GLIBC 2.27-3ubuntu1) 2.27
```
